crates/turborepo-lib/src/task_graph/visitor.rs

```diff
@@ -0,0 +1,108 @@
+use std::sync::{Arc, OnceLock};
+
+use futures::{stream::FuturesUnordered, StreamExt};
+use regex::Regex;
+use tokio::sync::mpsc;
+
+use crate::{
+    engine::{Engine, ExecutionOptions},
+    opts::Opts,
+    package_graph::{PackageGraph, WorkspaceName},
+    run::task_id::{self, TaskId},
+};
+
+// This holds the whole world
+pub struct Visitor<'a> {
+    package_graph: Arc<PackageGraph>,
+    opts: &'a Opts<'a>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("cannot find package {package_name} for task {task_id}")]
+    MissingPackage {
+        package_name: WorkspaceName,
+        task_id: TaskId<'static>,
+    },
+    #[error(
+        "root task {task_name} ({command}) looks like it invokes turbo and might cause a loop"
+    )]
+    RecursiveTurbo { task_name: String, command: String },
+    #[error("Could not find definition for task")]
+    MissingDefinition,
+    #[error("error while executing engine: {0}")]
+    Engine(#[from] crate::engine::ExecuteError),
+}
+
+impl<'a> Visitor<'a> {
+    pub fn new(package_graph: Arc<PackageGraph>, opts: &'a Opts) -> Self {
+        Self {
+            package_graph,
+            opts,
+        }
+    }
+
+    pub async fn visit(&self, engine: Arc<Engine>) -> Result<(), Error> {
+        let concurrency = self.opts.run_opts.concurrency as usize;
+        let (node_sender, mut node_stream) = mpsc::channel(concurrency);
+
+        let engine_handle = {
```

Comment:
We need the engine to execute in another task so it can send over the tasks via the node stream.